proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,566 @@
+millilamports per requested CU as the PRAW fees. Instead of having separate
+instructions for each compute budget parameter we will have a single instruction
+that will set all the compute budget parameters. Compute budget parameters could
+be considered as transaction header that sets some parameters for the
+transaction. We create a new instruction that takes a vector of parameters. If
+similar transaction header parameters appear multiple times then we just take
+the first one and ignore the other occurrences. In case of multiple transaction
+headers in the transaction, we take the first occurrence and ignore the others.
+
+```
+enum TransactionHeaderParameter {
+  // existing compute budget parameters
+  SetComputeUnitLimit(u32),
+  SetComputeUnitPrice(u64),
+  SetLoadedAccountsDataSizeLimit(u32),
+  // Limit on maximum of PRAW fees to be paid
+  // Fees is set in milli lamports per CU
+  SetProgramRebatableAccountWriteFees(u64),
+}
+
+enum TransactionHeaderInstruction {
+  parameter: Vec<TransactionHeaderParameter>
+}
+```
+
+When the cluster receives a new transaction, with `TransactionHeaderParameter`
+instruction is decoded to get the `SetProgramRebatableAccountWriteFees` and then
+calculates the total fees required for the transaction message. Then we verify
+that the fee-payer has a minimum balance of: `per-transaction base fees` +
+`prioritization fees` +
+`Millilamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+Where PRAW fees is :
+`Millilamports per CU set by SetProgramRebatableAccountWriteFees` *
+`requested CUs set by SetComputeUnitLimit`.
+
+If the payer does not have enough balance, the transaction is not scheduled and
+fails with an error `InsufficientFunds`. Consider this case the same as if the
+payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+PRAW fees. For all accounts with PRAW fees, the fees paid should be greater than
+the required fees. In case of overpayment, the difference is stored in a
+variable and will be paid back to the payer in any case. In case the PRAW fees
+are insufficiently paid or not paid then the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The PRAW fees minus rebates and minus overpayment are transferred to the
+respective accounts lamport balance, from where the owner can collect them and
+then the rest would be transferred back to the payer.
+
+### Syscall to change PRAW fees
+
+PRAW fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+A new syscall will be added to update the PRAW fees:
+
+```
+fn update_program_rebatable_account_write_fees(
+  account: Pubkey, 
+  milli_lamports_per_cu: u32
```

Comment:
Yeah we think u32 is enough in this case, if it is millilamports per requested CUs.
The thing is Account structure in solana barely have 3 more bytes in it, if we do not want to take the route to update the deprecated rent epoch and use PRAW fees instead of it.
This will also save some space on networking level.
But if you prefer u64 then we can change it to microlamports instead of milli lamports and then move also the decimals to set maximum PRAW fees that can be set on an account.

Reply:
I think it's ok. Not using word size just made my eyebrows raise

Reply:
Changed to u64 and microlamports per requested CUs